File: Leetcode - Python/286_graph_wallsAndGates.py
# https://leetcode.com/problems/walls-and-gates/

import collections

class Solution:
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        """
        if not rooms:
            return

        rows, cols = len(rooms), len(rooms[0])
        dirs = [(-1,0),(0,-1),(1,0),(0,1)]
        queue = collections.deque()
        for r in range(rows):
            for c in range(cols):
                if rooms[r][c] == 0:
                    queue.append((r,c))

        while queue:
            r, c = queue.popleft()
            distance = rooms[r][c] + 1
            for dx, dy in dirs:
                x, y = r + dx, c + dy
                if 0 <= x < rows and 0 <= y < cols and rooms[x][y] == 2147483647:
                    rooms[x][y] = distance
                    queue.append((x,y))

File: Leetcode - Java/286_graph_wallsAndGates.java
// https://leetcode.com/problems/walls-and-gates/

class Pair {
    int x, y;
    
    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
    
    public int getX() {
        return this.x;
    }
    
    public int getY() {
        return this.y;
    }
}

class Solution {
    public void wallsAndGates(int[][] rooms) {
        int rows = rooms.length;
        if (rows == 0) return;
        int cols = rooms[0].length;
        
        Queue<Pair> canReach = new LinkedList();
        int[][] dirs = new int[][]{{1,0},{-1,0},{0,1},{0,-1}};
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (rooms[i][j] == 0) {
                    canReach.add(new Pair(i, j));
                }
            }
        }
        
        while (!canReach.isEmpty()) {
            int r = canReach.peek().getX(), c = canReach.peek().getY();
            canReach.poll();
            
            for (int[] dir : dirs) {
                int x = r + dir[0], y = c + dir[1];
                if (x < 0 || x >= rows || y < 0 || y >= cols || rooms[x][y] <= rooms[r][c] + 1)
                    continue;
                rooms[x][y] = rooms[r][c] + 1;
                canReach.add(new Pair(x,y));
            }
        }
    }
}